File: src/utils/config_loader.py
import json
import os
from pathlib import Path

def load_config(local_config_path="config.json"):
    """
    Loads a global config.json from the project root and merges it with a 
    local config.json. Local parameters override global ones.
    """
    # Find project root (assumed to be 2 levels up from src/utils)
    utils_dir = Path(os.path.dirname(os.path.abspath(__file__)))
    project_root = utils_dir.parent.parent
    
    global_config_path = project_root / "config.json"
    config = {"paths": {}, "parameters": {}}

    # Load global config
    if global_config_path.exists():
        with open(global_config_path, 'r', encoding='utf-8') as f:
            global_config = json.load(f)
            for k, v in global_config.get("paths", {}).items():
                # Make paths absolute relative to project root
                config["paths"][k] = str(project_root / v)
            config["parameters"].update(global_config.get("parameters", {}))

    # Load local config
    search_dir = Path.cwd()
    local_found = False
    for _ in range(3):
        possible_path = search_dir / local_config_path
        if possible_path.exists() and possible_path.resolve() != global_config_path.resolve():
            with open(possible_path, 'r', encoding='utf-8') as f:
                local_config = json.load(f)
                if "paths" in local_config:
                    for k, v in local_config["paths"].items():
                        # If local path is relative, make it absolute from project root
                        # or keep it if it's already absolute
                        p = Path(v)
                        if not p.is_absolute():
                            config["paths"][k] = str(project_root / v)
                        else:
                            config["paths"][k] = v
                if "parameters" in local_config:
                    config["parameters"].update(local_config["parameters"])
                local_found = True
            break
        search_dir = search_dir.parent

    if not local_found and not global_config_path.exists():
         raise FileNotFoundError(f"Neither global config nor local config {local_config_path} found.")

    return config

File: src/utils/config_loader.jl
using JSON

function load_config(local_config_name="config.json")
    # Project root is assumed to be 2 levels up from src/utils/config_loader.jl
    utils_dir = dirname(abspath(@__FILE__))
    project_root = dirname(dirname(utils_dir))
    global_config_path = joinpath(project_root, "config.json")
    
    config = Dict("paths" => Dict(), "parameters" => Dict())

    # Load global config
    if isfile(global_config_path)
        global_config = JSON.parsefile(global_config_path)
        if haskey(global_config, "paths")
            for (k, v) in global_config["paths"]
                config["paths"][k] = abspath(joinpath(project_root, v))
            end
        end
        if haskey(global_config, "parameters")
            merge!(config["parameters"], global_config["parameters"])
        end
    end

    # Load local config
    current_dir = pwd()
    local_found = false
    for _ in 1:3
        possible_path = joinpath(current_dir, local_config_name)
        if isfile(possible_path) && abspath(possible_path) != abspath(global_config_path)
            local_config = JSON.parsefile(possible_path)
            if haskey(local_config, "paths")
                for (k, v) in local_config["paths"]
                    if isabspath(v)
                        config["paths"][k] = v
                    else
                        config["paths"][k] = abspath(joinpath(project_root, v))
                    end
                end
            end
            if haskey(local_config, "parameters")
                merge!(config["parameters"], local_config["parameters"])
            end
            local_found = true
            break
        end
        current_dir = dirname(current_dir)
    end

    if !local_found && !isfile(global_config_path)
        error("Neither global config nor local config $local_config_name found.")
    end

    return config
end
